### ISSUE TITLE
Use of OCRESTRA API end-points break backwards compatibility in PharmacoGx?

### ISSUE THREAD

**Issue by ChristopherEeles** [bhklab/ORCESTRA#22]
I just noticed this right now while trying to respond to a user issue for `PharmacoGx`.

When we migrated the `availablePSets` and `downloadPSet` functions to query the ORCESTRA API we started forcing all package users since this update to use the most recent `PharmacoSet` objects. This may break backwards compatibility if, for example, we migrate the data in the `@molecularProfilesSlot` to a `MultiAssayExperiment`. It may even break for new Bioconductor releases if there is a change in the object serialization in R from one release to another (e.g., as was done in version 3.5 -> 3.6), preventing users from loading `PharmacoSet`s without updating their version of R.

Since all the previously released `PharmacoSet`s are available on Zenodo, would it be possible to version the API URL such that we ensure users of previous `PharmacoGx` releases download the appropriate `PharmacoSet`s? Then we can update the URL for each release to prevent breaking backwards compatibility. That way users get the appropriate version of the PharmacoSet for the BiocManager version.

Let me know your thoughts on this?

Chris

**Comment by mnakano** [comment]
Hi Chris,

Yes, this can be implemented at the API endpoint. The following would be the expected behaviour of the new feature, if I understand correctly:

1. The current endpoint URL to get available PSets is https://www.orcestra.ca/psets/[canonical or available]
2. The modified version will have an additional 'version' parameter: https://www.orcestra.ca/psets/[canonical or available]/[version]
3. The API would use the version parameter to build a download URL for each PSet that corresponds to the version.

For the version parameter, would it be the PharmacoGx version? 

Minoru

**Comment by ChristopherEeles** [comment]
Hi Minoru,

Yes the solution you have outlined will work.

We can probably tie it to the Bioconductor releases in Spring and Fall. I guess it doesn't really matter if we use the PharmacoGx or Bioconductor release version, but my first instinct is to use the Bioconductor version since this is what BiocManager uses to determine the correct package to install/load.

For this release we would need a version for 3.13 and 3.14 (release and development, respectively). Because the spring release builds both branches with the same version of R (4.1.0), the two version could point to the same page. For the fall release, we will need two separate pages, since the development branch will start using R 4.2.0.

I guess the next question is do we want to implement this retroactively? I have a meeting with Petr this afternoon so I can talk to him then to get his thoughts.

Best,
Chris

**Comment by mnakano** [comment]
Hi Chris,

The latest deployment includes changes to handle multiple DOIs/download links for one data object depending on BioConductor release versions as follows:

1. The API end point URL can be used as it is, for example `/api/psets/canonical`.
2. It can have `version` query where the BioConductor release version would be passed: `/api/psets/canonical?version=3.13`.
3. The current default (and only) version is set to 3.13.
4. If no version is passed, the API returns data objects with DOI and download link for the current default version.

**Comment by ChristopherEeles** [comment]
Hey Minoru, 

I guess one complexity will be ensure the API still works when new releases occur. I think I can handle that in the R code right now by using a `tryCatch` block and decrementing the version until it find a match from the API.

As long as that works this should resolve this issue.

**Comment by ChristopherEeles** [comment]
No that is probably fine. The default will always be the highest available version right? I can just fall back to the default if I don't get results from my first call.

**Comment by mnakano** [comment]
Yes, it's intended to always be the highest version. The default value is controlled by the app's environment variable, so when we decide to change the default version value, all we need is to change the app's configuration value on Azure, and everything should work as is, provided that we have updated the database by adding the DOI and download link of the corresponding new version of the data objects.
